src/module/iotjs_module_httpparser.c

```diff
@@ -0,0 +1,554 @@
+  int i;
+  for (i = 0; i < _this->n_values; i++) {
+    iotjs_jval_t f = iotjs_jval_create_string(&_this->fields[i]);
+    iotjs_jval_t v = iotjs_jval_create_string(&_this->values[i]);
+    iotjs_jval_set_property_by_index(&jheader, i * 2, &f);
+    iotjs_jval_set_property_by_index(&jheader, i * 2 + 1, &v);
+    iotjs_jval_destroy(&f);
+    iotjs_jval_destroy(&v);
+  }
+  return jheader;
+}
+
+
+void iotjs_httpparserwrap_flush(THIS) {
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_httpparserwrap_t, httpparserwrap);
+
+  const iotjs_jval_t* jobj = iotjs_httpparserwrap_jobject(httpparserwrap);
+  iotjs_jval_t func = iotjs_jval_get_property(jobj, "OnHeaders");
+  IOTJS_ASSERT(iotjs_jval_is_function(&func));
+
+  iotjs_jargs_t argv = iotjs_jargs_create(2);
+  iotjs_jval_t jheader = iotjs_httpparserwrap_make_header(httpparserwrap);
+  iotjs_jargs_append_jval(&argv, &jheader);
+  iotjs_jval_destroy(&jheader);
+  if (_this->parser.type == HTTP_REQUEST &&
+      !iotjs_string_is_empty(&_this->url)) {
+    iotjs_jargs_append_string(&argv, &_this->url);
+  }
+
+  iotjs_make_callback(&func, jobj, &argv);
+
+  iotjs_string_make_empty(&_this->url);
+  iotjs_jargs_destroy(&argv);
+  iotjs_jval_destroy(&func);
+  _this->flushed = true;
+}
+
+
+void iotjs_httpparserwrap_set_buf(THIS, iotjs_jval_t* jbuf, char* buf, int sz) {
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_httpparserwrap_t, httpparserwrap);
+  _this->cur_jbuf = jbuf;
+  _this->cur_buf = buf;
+  _this->cur_buf_len = sz;
+}
+
+
+iotjs_jval_t* iotjs_httpparserwrap_jobject(THIS) {
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_httpparserwrap_t, httpparserwrap);
+
+  return iotjs_jobjectwrap_jobject(&_this->jobjectwrap);
+}
+
+
+http_parser* iotjs_httpparserwrap_parser(THIS) {
+  IOTJS_VALIDATED_STRUCT_METHOD(iotjs_httpparserwrap_t, httpparserwrap);
+  return &_this->parser;
+}
+
+
+#undef THIS
+
+
+#define HTTP_CB(name)                                                          \
+int name(http_parser* parser) {                                                \
+  iotjs_httpparserwrap_t* container = (iotjs_httpparserwrap_t*)(parser->data); \
+  return iotjs_httpparserwrap_##name(container);                               \
+}                                                                              \
+
+
+#define HTTP_DATA_CB(name)                                                     \
+int name(http_parser* parser, const char* at, size_t length) {                 \
+  iotjs_httpparserwrap_t* container = (iotjs_httpparserwrap_t*)(parser->data); \
+  return iotjs_httpparserwrap_##name(container, at, length);                   \
+}                                                                              \
+
+
+//http-parser callbacks
+HTTP_CB( on_message_begin );
+HTTP_DATA_CB( on_url );
+HTTP_DATA_CB( on_status );
+HTTP_DATA_CB( on_header_field );
+HTTP_DATA_CB( on_header_value );
+HTTP_CB( on_headers_complete );
+HTTP_DATA_CB( on_body );
+HTTP_CB( on_message_complete );
+
```

Comment:
Please check coding style guideline. We have a rule - ``Do not insert spaces after open paren and before close paren.``